File: digicomply/digicomply/report/trn_health_report/trn_health_report.py
# Copyright (c) 2024, DigiComply and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import now_datetime, date_diff, getdate


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "fieldname": "name",
            "label": _("TRN"),
            "fieldtype": "Link",
            "options": "TRN Registry",
            "width": 120
        },
        {
            "fieldname": "entity_name",
            "label": _("Entity Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "company",
            "label": _("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "width": 180
        },
        {
            "fieldname": "validation_status",
            "label": _("Validation Status"),
            "fieldtype": "Data",
            "width": 140
        },
        {
            "fieldname": "last_validated",
            "label": _("Last Validated"),
            "fieldtype": "Datetime",
            "width": 160
        },
        {
            "fieldname": "is_primary",
            "label": _("Is Primary"),
            "fieldtype": "Check",
            "width": 90
        },
        {
            "fieldname": "days_since_validation",
            "label": _("Days Since Validation"),
            "fieldtype": "Int",
            "width": 150
        }
    ]


def get_data(filters):
    conditions = get_conditions(filters)

    data = frappe.db.sql("""
        SELECT
            name,
            trn,
            entity_name,
            company,
            validation_status,
            last_validated,
            is_primary
        FROM `tabTRN Registry`
        WHERE is_active = 1
        {conditions}
        ORDER BY
            CASE validation_status
                WHEN 'Invalid' THEN 1
                WHEN 'Expired' THEN 2
                WHEN 'Not Validated' THEN 3
                WHEN 'Pending Verification' THEN 4
                WHEN 'Valid' THEN 5
                ELSE 6
            END,
            last_validated ASC
    """.format(conditions=conditions), filters, as_dict=True)

    # Calculate days since validation
    current_date = getdate(now_datetime())
    for row in data:
        if row.last_validated:
            row["days_since_validation"] = date_diff(current_date, getdate(row.last_validated))
        else:
            row["days_since_validation"] = None

    return data


def get_conditions(filters):
    conditions = []

    if filters.get("company"):
        conditions.append("AND company = %(company)s")

    if filters.get("validation_status"):
        conditions.append("AND validation_status = %(validation_status)s")

    if filters.get("show_only_invalid"):
        conditions.append("AND validation_status IN ('Invalid', 'Expired', 'Not Validated')")

    return " ".join(conditions)

File: digicomply/digicomply/report/trn_health_report/trn_health_report.js
// Copyright (c) 2024, DigiComply and contributors
// For license information, please see license.txt

frappe.query_reports["TRN Health Report"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "width": 180
        },
        {
            "fieldname": "validation_status",
            "label": __("Validation Status"),
            "fieldtype": "Select",
            "options": "\nNot Validated\nValid\nInvalid\nExpired\nPending Verification",
            "width": 140
        },
        {
            "fieldname": "show_only_invalid",
            "label": __("Show Only Invalid"),
            "fieldtype": "Check",
            "default": 0
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "validation_status" && data) {
            const status = data.validation_status;
            let color = "";

            switch (status) {
                case "Valid":
                    color = "#28a745";  // Green
                    break;
                case "Invalid":
                    color = "#dc3545";  // Red
                    break;
                case "Expired":
                    color = "#fd7e14";  // Orange
                    break;
                case "Not Validated":
                    color = "#6c757d";  // Gray
                    break;
                case "Pending Verification":
                    color = "#a404e4";  // Purple (brand color)
                    break;
                default:
                    color = "#6c757d";
            }

            value = `<span style="color: ${color}; font-weight: 600;">${status}</span>`;
        }

        if (column.fieldname === "days_since_validation" && data) {
            const days = data.days_since_validation;
            if (days !== null && days !== undefined) {
                let color = "";
                if (days > 90) {
                    color = "#dc3545";  // Red - overdue
                } else if (days > 60) {
                    color = "#fd7e14";  // Orange - warning
                } else {
                    color = "#28a745";  // Green - recent
                }
                value = `<span style="color: ${color}; font-weight: 500;">${days}</span>`;
            }
        }

        return value;
    }
};
